### ISSUE TITLE
可以在MAC上跑吗?

### ISSUE THREAD

**Issue by taozhiyuai** [jingyaogong/minimind#6]
可以在MAC上跑吗?

**Comment by jingyaogong** [comment]
训练框架基本上是原生PyTorch

CPU完全支持训练

但是PyTorch对MAC的GPU支持貌似只对M1芯片及以上有效，具体如何启用可以参考：

[how-to-enable-gpu-support-for-tensorflow-or-pytorch-on-macos](https://medium.com/bluetuple-ai/how-to-enable-gpu-support-for-tensorflow-or-pytorch-on-macos-4aaaad057e74)

[mac中pytorch使用](https://blog.csdn.net/weixin_42802447/article/details/128210650?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522272D728C-A32A-4E00-8AB0-1C8ECBEE4E6B%2522%252C%2522scm%2522%253A%252220140713.130102334.pc%255Fall.%2522%257D&request_id=272D728C-A32A-4E00-8AB0-1C8ECBEE4E6B&biz_id=0&utm_medium=distribute.pc_search_result.none-task-blog-2~all~first_rank_ecpm_v1~times_rank-19-128210650-null-null.142^v100^pc_search_result_base6&utm_term=torch%20mac&spm=1018.2226.3001.4187)

核心是把cuda换成mps
> MPS 是 Apple 提供的用于 macOS 和 iOS 设备的高性能计算库，专门用于 Apple 的 GPU 上的加速计算。
> CUDA 是 NVIDIA 提供的并行计算平台和 API 模型，专门用于 NVIDIA GPU 上的高性能计算。
```
device = torch.device("mps")
model = model.to(device)
```

BTW: 本人没有MAC设备供测试，不好意思~

**Comment by thinkhy** [comment]
可以尝试以下代码（minimind/scripts/web_demo.py):

    # Move model to the appropriate device with error handling
    try:
        model = model.eval().to(device)
    except Exception as e:
        if 'MPS' in str(e):
            print("Warning: MPS error encountered. Falling back to CPU.")
            global device
            device = "cpu"
            model = model.eval().to(device)
        else:
            raise e

**Comment by wisepmlin** [comment]
Apple M2 Max，执行一条后471
.venv(base) XXXXXX % python train_pretrain.py --device mps
LLM总参数量：25.830 百万
Epoch:[1/1](https://github.com/jingyaogong/minimind/issues/0/44160) loss:8.932 lr:0.000550000000 epoch_Time:6807.0min:
Epoch:https://github.com/jingyaogong/minimind/issues/100/44160 loss:6.404 lr:0.000549993674 epoch_Time:471.0min:

**Comment by wisepmlin** [comment]
<img width="613" alt="Image" src="https://github.com/user-attachments/assets/6aa65c05-a40c-4e64-b756-d522a42b009c" />
GPU运行
